0003-I-dapr-compose.md

```diff
@@ -0,0 +1,294 @@
+- if `env` field is present only in a particular `apps` section, any shell environment variables, `env` maps from `common` section and the `env` map for the current app will be passed to both the `app` and `daprd`.
+- The more specific `env` key-value pairs will override the lesser specific ones i.e. `apps` section specific `env` key-value pairs will override the key-value pairs from the `common` section which will override the passed in shell environment variables.
+
+
+For each app in the `apps` section, the `resources_dir` and `config_file` values will be resolved in the following order:
+- If `reosurce_dir` and/or `config_file` fields are given for any `apps[i]` configuration use that value as is.
+- If not, check for `.dapr` folder in the **`apps[i].app_dir`** folder. If found, use the `.dapr/resources` folder for configuring the resources and `.dapr/config.yaml` for the `daprd` configuration file(argument `--config-file` in `daprd`).
+- If not, check if a `resources_dir` and/or `config_file` fields are defined in the `common` section of the compose configuration file. If so use those values for those fields.
+- If not, default to `~/.dapr/resources/` for `resources_dir` and `~/.dapr/config.yaml` for `config_file` values.
+
+
+## Proposed command format
+
+Given the configuration file defined above, there should be a way to use the configuration file and run the differnt commands and `daprd` with the configuration given in the file.
+
+For this there will be a `compose` command that will be defined in the `dapr` CLI.
+
+### Constraints
+
+`compose` should mimic what `run` does for the most part. The defaults should not be different from `run` command with any new configuration options having sane documented least permissions based defaults.
+
+### Maintaining state and interaction flow
+A CLI command should be stateless unless its command is run as a foreground process and that command can store state in memory for the duration of the run.
+
+When multiple applications need to be run, monitored for crash(error exit) and restarted, state of the applications and associated other processes need to be maintained.
+
+#### How `dapr` currently maintains state across commands
+
+For getting the list of `daprd` processes and apps and then being able to stop them, `dapr` CLI is currently making use of the the metadata API that `daprd` provides directly.
+
+For example, on executing `dapr run`, there will always be an associated `daprd` along with any application that is run for that command. The PID for the CLI process, and the app command associated with it are put as extended metadata in the associated `daprd` process.
+
+When `dapr list` is run, it iterates through the OS processes list, gets all the listed `daprd` processes, queries their metadata API for getting details on the associated CLI PID, and APP command.
```

Comment:
ps -aef | grep <pid> with pid as the process id of the daemon, gives the list of processes whose parent is the daemon process. This would be a good start but would require to remember the pid of the daemon process. Can this be returned as a token or a id for every start?
P.S : This is very specific to Linux and I am hoping there should be an equivalent for each OS

Reply:
We are not running `ps` command since as you mentioned it is Linux specific, but we do use a library which returns all running processes.